contrib/99designs/gqlgen/tracer.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2016-2019 Datadog, Inc.
+
+// Package gqlgen contains an implementation of a gqlgen tracer, and functions to construct and configure the tracer.
+// The tracer can be passed to the gqlgen handler (see package github.com/99designs/gqlgen/handler)
+package gqlgen
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"github.com/99designs/gqlgen/graphql"
+)
+
+const (
+	tagResolverObject = "resolver.object"
+	tagResolverField  = "resolver.field"
+)
+
+const defaultResourceName = "gqlgen"
+
+type gqlTracer struct {
+	cfg config
+}
+
+// New creates an a graphql.Tracer instance that can be passed to a gqlgen handler.
+// Options can be passed in for further configuration.
+func New(opts ...Option) graphql.Tracer {
+	var t gqlTracer
+	defaults(&t.cfg)
+	for _, o := range opts {
+		o(&t.cfg)
+	}
+	return &t
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) StartOperationParsing(ctx context.Context) context.Context {
+	// not implemented
+	return ctx
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) EndOperationParsing(ctx context.Context) {
+	// not implemented
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) StartOperationValidation(ctx context.Context) context.Context {
+	// not implemented
+	return ctx
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) EndOperationValidation(ctx context.Context) {
+	// not implemented
+}
+
+func (t *gqlTracer) StartOperationExecution(ctx context.Context) context.Context {
+	rctx := graphql.GetRequestContext(ctx)
+	name := defaultResourceName
+	if rctx != nil && rctx.OperationName != "" {
+		name = rctx.OperationName
+	}
+	opts := []ddtrace.StartSpanOption{
+		tracer.SpanType(ext.SpanTypeGraphql),
+		tracer.ResourceName(name),
+		tracer.ServiceName(t.cfg.serviceName),
+	}
+	if !math.IsNaN(t.cfg.analyticsRate) {
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, t.cfg.analyticsRate))
+	}
+	if s, ok := tracer.SpanFromContext(ctx); ok {
+		opts = append(opts, tracer.ChildOf(s.Context()))
+	}
+	_, ctx = tracer.StartSpanFromContext(ctx, name, opts...)
+	return ctx
+}
+
+func (t *gqlTracer) StartFieldExecution(ctx context.Context, field graphql.CollectedField) context.Context {
+	span, ctx := tracer.StartSpanFromContext(ctx, "Field_"+field.Name)
+	span.SetTag("field", field.Name)
+	return ctx
+}
+
+func (t *gqlTracer) StartFieldResolverExecution(ctx context.Context, rc *graphql.ResolverContext) context.Context {
+	// This is the span created in StartFieldExecution.
+	// StartFieldResolverExecution is called only once per StartFieldExecution, so we can add context to the
+	// span rather than starting a child span.
+	span, ok := tracer.SpanFromContext(ctx)
+	if !ok {
+		return ctx
+	}
+	span.SetTag(ext.SpanName, rc.Object+"_"+rc.Field.Name)
```

Comment:
I'm sorry but I might've made a mistake in reasoning here. This needs to be removed because we are setting the name when the span is created. Here, we should set the resource, but can it maybe not use an `_` as a separator, but maybe a `.` or a `/` ?

I don't think we need the tags if we set the resource. Users are able to fuzz search through the resources in App Analytics.